[mm] Do not write diffuser model to disk when convert_cache set to zero

lstein opened the pull request:
## Summary

This PR does two things:

1. The model conversion and caching logic has changed. Now, if the convert cache directory is disabled by setting `convert_cache: 0` in the invoke config file, the converted model is never written to disk, and the converted model is kept in the RAM cache only. 
2. It cleans up the code that loads a model into memory. Instead of taking the model path and repo_variant as individually named arguments, the `_load_model()` method now accepts the entire model config object. This provides greater flexibility to control the load process.
3. 
<!--A description of the changes in this PR. Include the kind of change (fix, feature, docs, etc), the "why" and the "how". Screenshots or videos are useful for frontend changes.-->

## Related Issues / Discussions

See periodic discord user complaints about the convert cache directory.

<!--WHEN APPLICABLE: List any related issues or discussions on github or discord. If this PR closes an issue, please use the "Closes #1234" format, so that the issue will be automatically closed when the PR merges.-->

## QA Instructions

Try loading each type of model.

<!--WHEN APPLICABLE: Describe how we can test the changes in this PR.-->

## Merge Plan

Merge when approved.

<!--WHEN APPLICABLE: Large PRs, or PRs that touch sensitive things like DB schemas, may need some care when merging. For example, a careful rebase by the change author, timing to not interfere with a pending release, or a message to contributors on discord after merging.-->

## Checklist

- [X] _The PR has a short but descriptive title, suitable for a changelog_
- [X] _Tests added / updated (if applicable)_
- [X] _Documentation added / updated (if applicable)_

psychedelicious commented on invokeai/app/services/model_install/model_install_default.py:
```diff
@@ -501,7 +501,7 @@ def _scan_for_missing_models(self) -> list[AnyModelConfig]:
         missing_models: list[AnyModelConfig] = []
         for x in self.record_store.all_models():
-            if not Path(x.path).resolve().exists():
+            if not (self._app_config.models_path / x.path).resolve().exists():
```
Why do we need to change this if the model paths are all absolute?

lstein replied:
At least on my system, there are still lots of model paths that are relative and the system complains horribly about missing models at startup time.

psychedelicious commented on invokeai/backend/model_manager/load/model_loaders/ip_adapter.py:
```diff
@@ -24,13 +25,13 @@ class IPAdapterInvokeAILoader(ModelLoader):
             raise ValueError("There are no submodels in an IP-Adapter model.")
-        model = build_ip_adapter(
+        model_path = Path(config.path)
+        model: RawModel = build_ip_adapter(
```
I think the annotation here of `: RawModel` is unnecessary

lstein replied:
Mypy complains of a type mismatch on the return statement unless it is there.

psychedelicious replied:
Hmm, my editor is fine with it, but it uses pyright instead of mypy.

Is `model` a `RawModel`? It looks like the `build_ip_adapter` function returns one of a few different `IPAdapter` classes.

lstein replied:
`build_ip_adapter` returns either an `IPAdapter` or an `IPAdapterPlus`. `IPAdapter` is a subclass of `RawModel`, and `IPAdapterPlus` is a subclass of `IPAdapter`. So it works. Do you want a change to the more evolved class or union?

psychedelicious commented on invokeai/backend/model_manager/load/model_loaders/generic_diffusers.py:
```diff
@@ -28,14 +28,15 @@ class GenericDiffusersLoader(ModelLoader):
         if submodel_type is not None:
             raise Exception(f"There are no submodels in models of type {model_class}")
-        variant = model_variant.value if model_variant else None
+        repo_variant = getattr(config, "repo_variant", None)
```
There are a couple spots where we could use `isinstance` instead of `getattr`:

```py
repo_variant = config.repo_variant if isinstance(config, DiffusersConfigBase) else None
```

Personally I prefer `isinstance` for clarity but I don't feel too strongly about it.

lstein replied:
I don't mind either way, but have changed to `isinstance`.

psychedelicious commented on invokeai/backend/model_manager/load/load_default.py:
```diff
@@ -123,15 +104,39 @@ def get_size_fs(
+
+        return loaded_model
+
+    def _needs_conversion(self, config: AnyModelConfig, model_path: Path, dest_path: Path) -> bool:
+        return False
```
What is this doing?

lstein replied:
Setting a default for the `_needs_conversion()` method. Most models don't support conversion. The few that do override this method and have logic to determine whether the conversion needs to be performed (e.g. looks in the cache directory to see if the converted model is already cached, has an appropriate modification date, etc).

psychedelicious commented on invokeai/app/api/routers/model_manager.py:
```diff
@@ -630,7 +630,13 @@ async def convert_model(
     # loading the model will convert it into a cached diffusers file
-    model_manager.load.load_model(model_config, submodel_type=SubModelType.Scheduler)
+    try:
+        cc_size = loader._convert_cache.max_size
+        if cc_size == 0:  # temporary set the convert cache to a positive number so that cached model is written
+            loader._convert_cache.max_size = 1.0
+        loader.load_model(model_config, submodel_type=SubModelType.Scheduler)
+    finally:
+        loader._convert_cache.max_size = cc_size
```
`_convert_cache` isn't on the ABC so accessing it is unsafe

lstein replied:
Fair enough. Fixed.

psychedelicious replied:
Not safe to assign to an unknown attr: `loader._convert_cache.max_size = 1.0`

psychedelicious commented:
Nice, this will be very welcome for some users who are averse to the disk usage.

Do we expect measurable performance improvements when the cache is disabled?

lstein commented:
> Nice, this will be very welcome for some users who are averse to the disk usage.
> 
> Do we expect measurable performance improvements when the cache is disabled?

There are two reasons that performance may be improved on the _first_ load of the model.  First, there's reduction of disk I/O. With the convert cache enabled, the checkpoint is read, converted, the converted model gets written to disk and then immediately read from disk. So the weights get written or read three times. With the cache size set to zero, the the converted model doesn't get written to disk at all, so there's just the checkpoint read. I don't notice the improvement myself, but I'm using an SSD and I imagine there will be a noticeable improvement on spinning disk.

Second, I activated the `skip_torch_weight_init()` optimization that was previously used when loading diffusers models, but not checkpoint models. This does seem to improve loading time.

Of course the second and subsequent times that the model file is loaded, it will be slower than loading the cached diffusers version because the conversion has to be done again and it does take a little while to complete.

psychedelicious commented:
> There are two reasons that performance may be improved on the first load of the model. First, there's reduction of disk I/O. With the convert cache enabled, the checkpoint is read, converted, the converted model gets written to disk and then immediately read from disk. So the weights get written or read three times. With the cache size set to zero, the the converted model doesn't get written to disk at all, so there's just the checkpoint read. I don't notice the improvement myself, but I'm using an SSD and I imagine there will be a noticeable improvement on spinning disk.

I'd expect 2 disk I/O operations with the cache enabled:
- Read checkpoint from disk
- Convert checkpoint in memory to diffusers
- Write diffusers weights to cache

Why do we need to read the converted diffusers weights from disk when we just converted it in memory?

lstein commented:
> > There are two reasons that performance may be improved on the first load of the model. First, there's reduction of disk I/O. With the convert cache enabled, the checkpoint is read, converted, the converted model gets written to disk and then immediately read from disk. So the weights get written or read three times. With the cache size set to zero, the the converted model doesn't get written to disk at all, so there's just the checkpoint read. I don't notice the improvement myself, but I'm using an SSD and I imagine there will be a noticeable improvement on spinning disk.
> 
> I'd expect 2 disk I/O operations with the cache enabled:
> 
> * Read checkpoint from disk
> * Convert checkpoint in memory to diffusers
> * Write diffusers weights to cache
> 
> Why do we need to read the converted diffusers weights from disk when we just converted it in memory?

It was originally written this way to load submodels. When just the `unet`, say, is requested, the diffusers conversion code actually converts all the components in memory and we then write the converted pipeline, with all its subdirectories, to disk. Then, as the code is currently written, just the requested subfolder is read back into memory and returned.

To make the zero cache directory work, I have to push all the converted submodels into the RAM cache after in-memory conversion, even if they weren't requested. Otherwise the code will convert the pipeline just to get the unet, convert it again to get the vae, and so forth.

Since I suspect you're about to ask for this. I'll just do the same thing when the disk cache is enabled as well, to avoid that extra set of reads.

psychedelicious commented:
Makes sense, thanks for explaining.

> Since I suspect you're about to ask for this.

😅

lstein commented:
> Makes sense, thanks for explaining.
> 
> > Since I suspect you're about to ask for this.
> 
> 😅

Done.
